### DISCUSSION TITLE
mark2 is not using console colors

### DISCUSSION THREAD

**Discussion by TomLewis** [gsand/mark2#141] [General]
Everything is black and white in the live console for Mark2, how can I get colors to show for colored items in the console? Like player names etc.

**Comment by RhvsW** [comment]
Terminal colors are stripped by default, see: `https://github.com/gsand/mark2/blob/master/mk2/resources/mark2.default.properties#L102`

**Comment by TomLewis** [comment]
How do I disable it to get a colored console again?

**Comment by gsand** [comment]
I honestly don't remember. I've been kinda busy lately and haven't had time to do mark2 updates.
:cry:

**Comment by TomLewis** [comment]
I would very much appreciate getting colours in the console, if you have some time @gsand could you please look into how I can get colors working in the console. Thank you.

**Comment by TomLewis** [comment]
Bump. 
Please can we get proper colors showing in terminal, this plane white is impossible to read.

**Comment by danthedaniel** [comment]
You can just comment on that line in the default properties, right?

**Comment by RhvsW** [comment]
mark2 doesn't parse the colors at all.

**Comment by TomLewis** [comment]
Bump; can we implement colors; this bash script is pre-made and needs to be merged somehow into Makr2, this fella is the one that makes the timing reports website and paperspigot.
https://aikar.co/2015/07/28/bash-colors-for-minecraft-shell-scripts/

**Comment by KasperFranz** [comment]
Has anyone been able to figure this out?

I can see from another issue that @Sir-Will got it working on a previous issue ( over a year ago) (the issue wasn't about missing colors but more about a colour not showing correct #35)

**Comment by Sir-Will** [comment]
I have this issue too, on few servers it shows colors on the others not, although they have the same plugins and version.

**Comment by KasperFranz** [comment]
hmm is it related to the os?

**Comment by Sir-Will** [comment]
@KasperFranz the servers are on the same box, so same ubuntu version.

**Comment by KasperFranz** [comment]
oh that is weird then o.0 

and guessing nearly same config?

**Comment by Sir-Will** [comment]
Yea, already looked for a difference, couldn't find one.

**Comment by KasperFranz** [comment]
that is weird, I have even tried with 1.8 and 1.9 servers which is also scrapping the colors (planning on using it for a 1.7 server)

**Comment by KasperFranz** [comment]
I have now used 2 hours trying to figure it out, but with my limited knowledge about python it is hard :P 

it seems that the unicodes are scrapped from the input to the colorize (which seems to be done before called the append_log function - that is how far i got with it before i couldn't understand anything about it :p

**Comment by mattboy9921** [comment]
It's been a while since much has been discussed here. I'm still experiencing this issue. When starting the server, I get a `WARN Disabling terminal, you're running in an unsupported environment.` I'm not sure if that has anything to do with it. Additionally, I've spent time testing various edits to the `java.cli.D.jline.terminal=jline.UnsupportedTerminal` line as well as testing different versions of Minecraft (Vanilla, Spigot and Paper). 

What's interesting to me is looking at [https://github.com/gsand/mark2/blob/master/mk2/user_client.py#L681](url) there is an entire `colorize` function that looks to be doing just what we want, colors in console. This project is quite large and being that I don't use Python every day, I've only really poked around a bit. I'm not sure why console is colorless despite there being code to support color. Hopefully this helps.

**Comment by TomLewis** [comment]
I would still like colors to be parsed in console too.

**Comment by mattboy9921** [comment]
After some more digging, the `WARN Disabling terminal, you're running in an unsupported environment.` message appears when the server isn't being run directly on the command line, so it disables a module that would have given it color. That message appears regardless of what I do with the properties files. 

I have been trying to trace my way through the program to figure out how mark2 starts the server. I figure there must be something hard coded somewhere that is telling the Java program that it is running in an unsupported terminal. So far I have come up blank though. 

(Edit)
So now I believe whatever mark2 is doing to start the terminal doesn't provide proper ANSI support. At least with Paper, that is checked here: https://github.com/PaperMC/Paper/blob/fb25dc17c62acce335346d5e96899257ae7230c7/Spigot-Server-Patches/0176-Use-TerminalConsoleAppender-for-console-improvements.patch#L212 and that same Java flag, `-Djline.terminal=jline.UnsupportedTerminal`, gets set anyway. So that means that however mark2 is handling the terminal triggers the Minecraft server to disable JLine which I assume handles the colors.

**Comment by Column01** [comment]
Did some digging and ANSI codes do work, but all player input that is sent to the console has color codes stripped before we receive it so we can't format it after the fact. As for the unsupported terminal flag, I believe that is mandatory as it will crash the server without it in some instances.

You can test it by sending a color-coded message using the say command from the console:
![image](https://user-images.githubusercontent.com/26526649/127082177-14c6e5ae-c64c-4524-98fc-87bc067c0b00.png)

I looked at what gets received when a line is printed on the console, and it has all the formatting stripped which would be why it doesn't get colorized. I'm not entirely sure how to set it up where formatting is passed on to mark2.

Regardless of this, I made some changes to the code to allow the use of color codes from the console sender in the form of ampersands (`&`) as long as the command is a `say` command. This should hopefully prevent errors but honestly not sure we even need to check that its a say command as I've never seen a command use `&<number>` in it

See commit: https://github.com/gsand/mark2/commit/a684890e08104204390add95959cb0dc660228af

**Comment by Column01** [comment]
Removing the unsupported terminal flag still yielded no colors so that's not related I don't think (at least on mark2's end)

**Comment by Column01** [comment]
Ok, interesting... I just tried on another server running Mohist 1.16.5 (forge-spigot similar to thermos) and the server saying something showed as colored in reply, but doing it as a player did not.
![image](https://user-images.githubusercontent.com/26526649/127085443-8e491f44-03ae-4c39-a9ba-fbf447b4226e.png)

**Comment by mattboy9921** [comment]
@Column01 Thank you for the commits and research on this! Glad to see someone else poking around lol. 

Now correct me if I'm wrong but is it not the case that the [Warning] and [Severe] messages should be colored? Or perhaps that is just a feature of certain consoles? I honestly rarely use anything other than Mark2, even just bare terminal so I am quite unaware if it is colored there or not.

**Comment by Column01** [comment]
> Now correct me if I'm wrong but is it not the case that the [Warning] and [Severe] messages should be colored?

I've not tested running these instances bare in a terminal, might have to try that. It's been forever since I've run a server through an actual terminal and not using mark2.

**Comment by Column01** [comment]
Again, on the mohist server it appears that Unicode color-coded strings that are from the server are colored in the console: 
![image](https://user-images.githubusercontent.com/26526649/127178920-9854ed4c-eb27-44de-a9e0-1cb69eccf840.png)
Definitely feel like this is a spigot thing where it disables color codes in the console from players and their names, probably to allow for less confusion with color-coded warnings/errors. Speaking of which, it should be possible for mark2 to automatically color code the level info with some minor changes.

**Comment by Column01** [comment]
What do you think about this? Does it look similar to what happens for errors from a standard console? The problem with this is it will not work for multi-line errors (for example a stack trace) afaik. Also considering just highlighting the level message
![image](https://user-images.githubusercontent.com/26526649/127191063-3cc231e6-71b7-4e3d-bc0b-9ee0aec2b293.png)

**Comment by mattboy9921** [comment]
@Column01 That does look pretty much like what I've seen in other consoles (keep in mind I am in the same boat as you, mainly use Mark2 so)

**Comment by Column01** [comment]
What about highlighting just the log level? Do you think that would look better? Also fixes not highlighting a full multiline error message, but still draws attention to issues.
![image](https://user-images.githubusercontent.com/26526649/127195755-d469ea0b-51e3-493c-9f2e-ca9c8cd05ee3.png)

**Comment by mattboy9921** [comment]
I think highlighting just the log level does make sense since you said stack traces don't get highlighted correctly but I honestly prefer the whole line to be highlighted. Perhaps it is possible to add a config option?

**Comment by Column01** [comment]
I don't think a config option for something like this makes sense considering level highlighting is the default behavior for the Minecraft server console. Come to think of it, the only errors that won't get highlighted in the multiline form are those that use `e.printStackTrace()` instead of properly logging it (since it will be in `RAW` or some other level instead of a proper log level)

For example:
![image](https://user-images.githubusercontent.com/26526649/127199442-62135c84-e684-41b5-b338-28aecb043a6e.png)

I suppose I'll just do full line highlighting then

**Comment by Column01** [comment]
Added the log level color changes in https://github.com/gsand/mark2/commit/15ece58e44f2f0f4e3735e61530122c71760d66d

**Comment by Column01** [comment]
My suspicions were correct, proper multiline stack trace printing works as intended.
![image](https://user-images.githubusercontent.com/26526649/127206637-bbed3a3b-969d-4a7c-a66a-604275ee1c7d.png)

**Comment by Column01** [comment]
Moved to a discussion, this isn't an issue with mark2 and is likely the server software stripping the color data as discussed.

**Comment by TomLewis** [comment]
I cant wait to have some sort of highlighting

**Reply by Column01** [reply]
The latest mark2 has highlighting for warnings and errors
